File: problems/typescript/2-two-pointers/0011_container_with_most_water/solution.ts
/**
 * Container With Most Water
 * https://leetcode.com/problems/container-with-most-water/
 * Time: O(n), Space: O(1)
 *
 * Two pointer approach - move pointer with smaller height
 * - Reviewed and resolved 01/10/2026
 */
export function maxArea(height: number[]): number {
	let left = 0;
	let right = height.length - 1;
	let maxArea = 0;

	while (left < right) {
		maxArea = Math.max(
			Math.min(height[left], height[right]) * (right - left),
			maxArea
		);

		if (height[left] <= height[right]) {
			left++;
		} else {
			right--;
		}
	}

	return maxArea;
}

File: problems/python/2-two-pointers/0011_container_with_most_water/solution.py
from typing import List

class Solution:
    """
    Time: O(n), Space: O(1)

    Two pointer approach - move pointer with smaller height
    - Reviewed 01/10/2026
    """
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_area = 0

        while left < right:
            current_area = (right - left) * min(height[left], height[right])
            max_area = max(max_area, current_area)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area